lib/spack/llnl/util/filesystem.py

```diff
@@ -1712,66 +1715,123 @@ def find(root, files, recursive=True):
-    # found in a key, and reconstructing the stable order later.
-    found_files = collections.defaultdict(list)
+@system_path_filter(arg_slice=slice(1))
+def find_max_depth(root, globs, max_depth: Optional[int] = None):
+    """Given a set of non-recursive glob file patterns, finds all
+    files matching those patterns up to a maximum specified depth.
 
-    # Make the path absolute to have os.walk also return an absolute path
-    root = os.path.abspath(root)
-    for path, _, list_files in os.walk(root):
-        for search_file in search_files:
-            matches = glob.glob(os.path.join(path, search_file))
-            matches = [os.path.join(path, x) for x in matches]
-            found_files[search_file].extend(matches)
+    If a directory has a name which matters an input pattern, it will
+    not be included in the results.
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    if max_depth is None:
+        max_depth = sys.maxsize
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
-    found_files = collections.defaultdict(list)
+    if isinstance(globs, str):
+        globs = [globs]
+    # Apply normcase to regular expressions and to the filenames:
+    # this respects case-sensitivity semantics of different OSes
+    # (e.g. file search is typically case-insensitive on Windows)
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
 
-    # Make the path absolute to have absolute path returned
+    if not os.path.isdir(root):
+        raise ValueError(f"{root} is not a directory")
+
+    # Note later calls to os.scandir etc. return abspaths if the
+    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
     root = os.path.abspath(root)
 
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
+    found_files = collections.defaultdict(list)
+
+    def _follow_link(p):
+        last = p
+        while os.path.islink(last):
+            last = resolve_link_target_relative_to_the_link(last)
+        return last
+
+    def _dir_id(p):
+        stat_result = os.stat(p, follow_symlinks=True)
+        # Note: on windows, st_ino is the file index and st_dev
+        # is the volume serial number. See
+        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
```

Comment:
We use this pattern in tons of places, I don't think it's worth documenting.

Reply:
> We use this pattern in tons of places, I don't think it's worth documenting.

I think this is essential as a comment unless somewhere mentions that `st_dev` is the volume serial number

This is not well explained in the python docs, and it could easily have been wrong in all the tons of places we use it (but that wouldn't have been very easy to detect because it would have only come up when you need to differentiate files that live on different volumes, which may not yet have been encountered for us on Windows)

> The thing that is worth commenting is

Good point, I'll add that too